gossipd/gossipd.c

```diff
@@ -2105,45 +2132,73 @@ static struct io_plan *getchannels_req(struct io_conn *conn,
 				       const u8 *msg)
 {
 	u8 *out;
-	const struct gossip_getchannels_entry **entries;
+	struct gossip_getchannels_entry *entries;
+	struct gossip_halfchannel_entry *half_entries;
+	const struct gossip_getchannels_entry **entryp;
 	struct chan *chan;
 	struct short_channel_id *scid;
 	struct node_id *source;
+	size_t n = 0, n_half = 0;
 
 	/* Note: scid is marked optional in gossip_wire.csv */
 	if (!fromwire_gossip_getchannels_request(msg, msg, &scid, &source))
 		master_badmsg(WIRE_GOSSIP_GETCHANNELS_REQUEST, msg);
 
-	entries = tal_arr(tmpctx, const struct gossip_getchannels_entry *, 0);
 	/* They can ask about a particular channel by short_channel_id */
 	if (scid) {
+		entries = tal_arr(tmpctx, struct gossip_getchannels_entry, 1);
+		half_entries = tal_arr(tmpctx, struct gossip_halfchannel_entry,
+				       tal_count(entries) * 2);
 		chan = get_channel(daemon->rstate, scid);
 		if (chan)
-			append_channel(daemon->rstate, &entries, chan, NULL);
+			add_channel_entry(daemon->rstate, &entries, &n,
+					  &half_entries, &n_half, chan, NULL);
 	} else if (source) {
 		struct node *s = get_node(daemon->rstate, source);
 		if (s) {
 			struct chan_map_iter i;
 			struct chan *c;
 
+			entries = tal_arr(tmpctx,
+					  struct gossip_getchannels_entry, 4);
+			half_entries = tal_arr(tmpctx,
+					       struct gossip_halfchannel_entry,
+					       tal_count(entries));
 			for (c = first_chan(s, &i); c; c = next_chan(s, &i)) {
-				append_channel(daemon->rstate,
-					       &entries, c, source);
+				add_channel_entry(daemon->rstate,
+						  &entries, &n,
+						  &half_entries, &n_half,
+						  c, source);
 			}
 		}
 	} else {
 		u64 idx;
 
+		/* Starting guess at 4 channels per node. */
+		entries = tal_arr(tmpctx,
+				  struct gossip_getchannels_entry,
+				  (daemon->rstate->nodes->raw.elems + 1) * 4);
```

Comment:
This now has enough magic in it that it might be worth it just to iterate through the map twice, wouldn't it? That way we'd get an exact count, could directly allocate the correct size and wouldn't need to overallocate and to re-allocate if our over-allocation was too small anyway.

Reply:
I actually needed to do more work here; by jamming the entire response into lightningd at once, it ends up trying to allocate 1GB for the JSON, which does not work on the rPi I have, crashing lightningd.

The patch to make it batch is fairly simple, has nice performance properties (we can process in parallel while gossipd is generating replies) and that supercedes this anyway.  I'll drop it.